python/ray/dashboard/modules/node/node_head.py

```diff
@@ -170,104 +166,121 @@ def get_internal_states(self):
+                overwrite=True,
+                namespace=ray_constants.KV_NAMESPACE_JOB,
+                timeout=GCS_RPC_TIMEOUT_SECONDS,
+            )
+        assert node["state"] in ["ALIVE", "DEAD"]
+        is_alive = node["state"] == "ALIVE"
+        # Prepare agents for alive node, and pop agents for dead node.
+        if is_alive:
+            if node_id not in DataSource.agents:
+                # Agent port is read from internal KV. Problem is it's not present when
+                # we receive this update; it's only present after agent.py starts
+                # listening. So we make an async task that periodically polls internal
+                # KV.
+                asyncio.create_task(self._update_agent(node_id))
+        else:
+            DataSource.agents.pop(node_id, None)
+            self._dead_node_queue.append(node_id)
+            if len(self._dead_node_queue) > node_consts.MAX_DEAD_NODES_TO_CACHE:
+                DataSource.nodes.pop(self._dead_node_queue.popleft(), None)
+        DataSource.nodes[node_id] = node
+
+    async def _update_agent(self, node_id):
+        """
+        Given a node, update the agent_port in DataSource.agents. Problem is it's not
+        present until agent.py starts, so we need to loop waiting for agent.py writes
+        its port to internal kv.
+        """
+        key = f"{dashboard_consts.DASHBOARD_AGENT_PORT_PREFIX}{node_id}".encode()
+        while True:
+            agent_port = await self._gcs_aio_client.internal_kv_get(
+                key,
+                namespace=ray_constants.KV_NAMESPACE_DASHBOARD,
+                timeout=GCS_RPC_TIMEOUT_SECONDS,
+            )
+            # The node may be dead already. Only update DataSource.agents if the node is
+            # still alive.
+            if DataSource.nodes.get(node_id, {}).get("state") != "ALIVE":
+                return
+            if agent_port:
+                DataSource.agents[node_id] = json.loads(agent_port)
+                return
+            await asyncio.sleep(node_consts.RAY_NODE_HEAD_AGENT_POLL_INTERVAL_S)
+
+    async def _update_nodes(self):
+        """
+        Subscribe to node updates and update the internal states. If the head node is
+        not registered after RAY_NODE_HEAD_HEAD_NODE_REGISTRATION_TIMEOUT, it logs a
+        warning only once.
+        """
+        warning_shown = False
+        async for node in self._subscribe_nodes():
+            await self._update_node(node)
+            if not self._head_node_registration_time_s:
+                # head node is not registered yet
                 if (
-                    head_node_not_registered
-                    and self._node_update_cnt * FREQUENTY_UPDATE_NODES_INTERVAL_SECONDS
-                    < FREQUENT_UPDATE_TIMEOUT_SECONDS
+                    not warning_shown
+                    and (time.time() - self._module_start_time)
+                    > RAY_NODE_HEAD_HEAD_NODE_REGISTRATION_TIMEOUT
```

Comment:
We can just do `RAY_NODE_HEAD_REGISTRATION_TIMEOUT`

Reply:
changed to RAY_DASHBOARD_HEAD_NODE_REGISTRATION_TIMEOUT